File: 07_stacks_and_queues/fish.py
def challenge(size, stack):
  while len(stack)>0:
    if stack[-1][1] == 0:
      stack.append((size, 0))
      return stack
    elif stack[-1][0] > size:
      return stack
    else:
      stack.pop()
  stack.append((size, 0))
  return stack

def solution(A, B):
  l = len(A)
  stack = []
  for i in xrange(l):
    size = A[i]
    way = B[i]
    if way == 0:
      stack = challenge(size, stack)
    else:
      stack.append((size, way))
  return len(stack)

File: 07_stacks_and_queues/fish.cpp
#include "vector"
#include "stack"

using namespace std;

int solution(vector<int> &a, vector<int> &b) {
  int ret = 0;
  int n = a.size();
  stack<int> stk;
  for (int i = 0; i < n; i++) {
    if (b[i] == 1) {
      stk.push(a[i]);
    }
    else if (stk.size() == 0) {
      ret++;
    }
    else {
      while (stk.size() > 0 && stk.top() < a[i]) {
        stk.pop();
      }
      if (stk.size() == 0) {
        ret++;
      }
    }
  }
  return ret + stk.size();
}
